internal/overlord/logstate/pumper.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	logReaderSize = 4 * 1024
+)
+
+// pumper is responsible for "pumping" the logs from the service ringbuffer
+// through to the log target client.
+type pumper struct {
+	// Metadata
+	service string
+	target  plan.LogTarget
+
+	iterator servicelog.Iterator
+	parser   *servicelog.Parser
+	client   logClient
+	stopCh   chan os.Signal
+}
+
+func newPumper(svc string, it servicelog.Iterator, tgt plan.LogTarget) (*pumper, error) {
+	client, err := newLogClient(svc, tgt)
+	if err != nil {
+		return nil, err
+	}
+
+	stopCh := make(chan os.Signal, 1)
```

Comment:
There are a number of reasons why we shouldn't be creating a stop signal on os sig's.

The best one for explanation is this pumper has a very good chance of stopping before the service and so we will potentially drop log message from the service as it shuts down.

Stopping of a pumper should be tied into the normal shut down of pebble.

Reply:
the whole point of this was to flush the buffer when Ctrl-C is pressed rather than dropping the messages

Reply:
If I understand what Tom is saying here, I agree. We should only handle OS signals in main at the very top level. Here we should be passing in a context or stop channel, and cancel/close that (in main, or passed down from main) when Ctrl-C is pressed.

Reply:
I'm thinking each service should notify the log manager when it stops, then the log manager can stop the relevant pumpers. Can't quite work out where the actual service stopping code is though